crates/matrix-sdk-base/src/sliding_sync.rs

```diff
@@ -0,0 +1,234 @@
+#[cfg(feature = "e2e-encryption")]
+use std::ops::Deref;
+
+use matrix_sdk_common::deserialized_responses::{
+    AmbiguityChanges, JoinedRoom, Rooms, SyncResponse,
+};
+use ruma::api::client::sync::sync_events::{v3, v4};
+#[cfg(feature = "e2e-encryption")]
+use ruma::UserId;
+
+use super::BaseClient;
+use crate::{
+    error::Result,
+    rooms::RoomType,
+    store::{ambiguity_map::AmbiguityCache, StateChanges},
+};
+
+impl BaseClient {
+    /// Process a response from a sliding sync call.
+    ///
+    /// # Arguments
+    ///
+    /// * `response` - The response that we received after a successful sliding
+    ///   sync.
+    pub async fn process_sliding_sync(&self, response: v4::Response) -> Result<SyncResponse> {
+        #[allow(unused_variables)]
+        let v4::Response {
+            // not applicable.
+            // next_batch,
+            rooms,
+            lists,
+            // FIXME: missing compared to v3::Response
+            //presence,
+            //account_data,
+            //to_device,
+            //device_lists,
+            //device_one_time_keys_count,
+            //device_unused_fallback_key_types,
```

Comment:
Is this `FIXME` relevant today? Can we fix it today? Is it also related to https://github.com/matrix-org/matrix-rust-sdk/issues/1014?

Reply:
still relevant, yes. This is done via extensions. I could clean this up, but felt more comfortable to keep the copy-pasted-code in so ensure we do them in a similar order ....

Reply:
:+1: